[2단계 - 로또 리팩토링] 줄리(한승아) 미션 제출합니다.

jbilee opened the pull request:
오리, 안녕하세요! :)

로또 미션 2단계에 대한 PR 올립니다.
1단계 피드백 적용과 함께 2단계의 3가지 요구사항은 아래와 같이 고민하며 수행했습니다. 리뷰해주셔서 감사합니다!

## 1. 핵심 비지니스 로직을 가지는 객체를 domain 패키지, UI 관련한 객체를 view 패키지에 구현한다.
UI와 관련된 객체(`InputView`, `OutputView`)는 이미 view 패키지에 구현했기 때문에 핵심 비즈니스 로직을 domain에 적절하게 구현했는지를 중점적으로 확인했습니다. 두가지 질문을 기반으로 개선할 부분을 찾아 리팩토링을 진행했습니다.

### 지금까지 구현한 도메인은 모두 핵심 비즈니스 로직만 가지고 있나?

#### 핵심 비즈니스 로직에 해당하는 기능:
```
- 로또를 발행 (사용자가 구매한 로또 및 당첨 로또)
- 로또 번호를 비교하여 일치 개수를 확인
- 당첨 번호와 보너스 번호의 일치 여부를 토대로 당첨 등수를 확인
- 당첨 등수에 따른 당첨 금액을 계산
- 당첨 금액과 구매 금액을 비교해 이율을 계산
```
#### 도메인별 체크리스트:
- ✅ Lotto 도메인
  - `hasNumber`(특정 번호가 로또 번호 중에 존재하는지를 확인)와 `findMatches`(특정 번호들이 로또 번호와 몇 개나 일치하는지를 확인) 메소드는 애플리케이션의 필수 기능 중 하나인 로또 번호 일치 여부와 개수를 확인하는 데에 필요하여 핵심 비즈니스 로직만 가지고 있는 것으로 생각됨
- ✅ WinningLotto 도메인
  - `findWinningTier` 메소드는 인자로 받은 로또 객체와 당첨 로또 객체의 번호를 비교해 어느 당첨 등수에 해당하는지를 판단하며, 이는 당첨 번호와 보너스 번호의 일치 여부를 토대로 당첨 등수를 확인하는 핵심 비즈니스 로직인 것으로 생각됨
- ❓ Scoreboard 도메인
  - `findWinningTiers` 메소드는 단순히 구매 로또 리스트와 당첨 로또를 인자로 받아 for문으로 `WinningLotto`의 `findWinningTier` 메소드만 실행시켜 결과로 반환되는 `WinningTier`를 모아주는 aggregator 역할만 하는 것인데, 반환하는 `WinningTier` 리스트는 이후에 당첨 금액과 이율을 계산할 때 필요하긴 하지만 행위 그 자체만으로 핵심 비즈니스 로직인지는 의심됨
- ✅ Vendor 도메인
  - `calculateLottoCount` 및 `issueLottos` 메소드는 로또 발행에 필요하며, `calculateProfit` 메소드는 이율 계산에 필요하기 때문에 핵심 비즈니스 로직만 갖고 있는 것으로 생각됨

#### 수정사항:
Scoreboard가 수행하는 액션은 단순히 다른 도메인에게 당첨 등수를 찾아달라는 메세지를 던져 그 결과물을 합치기만 하는 것이며, 이는 컨트롤러에게 더 적합한 역할이라고 생각해 Scoreboard 도메인을 삭제하고 Scoreboard의 메소드를 컨트롤러가 수행하도록 수정했습니다.

### 현재 컨트롤러에 도메인으로 뺄 수 있는 핵심 비즈니스 로직이 있나?
컨트롤러에는 뷰와 도메인 사이에서 정보를 전달해주는 코드만 있어서 핵심 비즈니스 로직이 없는 것으로 판단되어 따로 수정한 부분은 없습니다.

## 2. MVC 패턴 기반으로 리팩터링해 view 패키지의 객체가 domain 패키지 객체에 의존할 수 있지만, domain 패키지의 객체는 view 패키지 객체에 의존하지 않도록 구현한다.

MVC 패턴의 조건을 잘 맞췄는지 구현한 코드의 구조를 검토했습니다:
- ✅ view 패키지의 객체가 domain 패키지 객체에 의존할 수 있다
  - 제가 구현한 코드 내 `OutputView` 객체는 `Lotto` 객체에 의존하고 있습니다. 비즈니스 로직이 배출한 결과값을 UI에 출력하기 위해서는 domain 객체에게 출력할 값을 요청해야 하기 때문에, 이 부분에 대한 코드는 수정하지 않았습니다.
- ✅ domain 패키지의 객체는 view 패키지 객체에 의존하지 않는다
  - 현재 domain 패키지 내 객체들은 view 객체를 사용하지 않는다는 점 확인되어 코드는 수정하지 않았습니다.

## 3: 테스트 가능한 부분과 테스트하기 힘든 부분을 분리해 테스트 가능한 부분에 대해서만 단위 테스트를 진행한다.

이번 미션에서 랜덤성을 테스트할 수 있는 방법(seed값을 직접 주입해 Random API가 리턴하는 결과값을 조작, 또는 랜덤성이 들어있는 메소드를 override해 해당 메소드가 리턴하는 결과값을 조작)을 학습하였으나 제가 구현한 애플리케이션 내에서 랜덤성을 테스트해야 하는 이유는 무엇인가?에 대한 질문을 스스로 고민해봤을 때, 이번 미션에서는 달리 테스트해야 할 이유를 찾지 못했습니다. 핵심 비즈니스 로직 중에서 랜덤으로 리턴된 값이 필요한 경우는 로또를 생성하기 위해서인데, 해당 랜덤값들을 제가 조작해도 로또 생성에는 영향이 없어서 자바에 빌트인된 Random API를 override하면서까지 테스트할 필요성이 없다고 판단했습니다.

따라서 UI(뷰)를 제외하고, 컨트롤러와 도메인이 가진 메소드들만 단위 테스트가 작성되어 있습니다.

jbilee committed:
style(test): 테스트 코드 스타일 수정

작업 내용
- 각 테스트 코드에 개행 문자 추가

jbilee committed:
refactor(util): 난수 생성 관련 로직 리팩토링

작업 내용
- Set을 활용해 중복이 없는 숫자 리스트를 보장하도록 수정
- 매직넘버를 제거하고 외부로부터 생성할 숫자의 개수를 받도록 수정
- 숫자 정렬을 `RandomGenerator`가 수행하도록 수정

jbilee committed:
test(util): 난수 생성 관련 테스트 코드 수정

작업 내용
- 수정된 클래스 로직에 맞춰 테스트 수정

jbilee committed:
refactor(lotto): 로또 도메인에 오버로딩 생성자 추가

작업 내용
- `Lotto` 클래스 생성자에서 List 외 일반 배열도 매개변수로 받을 수 있도록 수정

jbilee committed:
test(lotto): 로또 테스트 코드 리팩토링

작업 내용
- `Lotto` 클래스 생성자 수정에 따라 불필요한 List 래퍼를 args 목록으로 대체
- 여러가지 상황을 검증하는 테스트를 세부적인 조건에 따라 분리
- 테스트명을 직관적으로 개선

jbilee committed:
style(constant): 숫자 표기법 개선

작업 내용
- 가독성을 위해 천 단위마다 `_` 구분자 추가

jbilee committed:
refactor(constant): 등수 열거형 내 로직 개선

작업 내용
- 직관적인 Stream 사용을 통해 `WinningTier` 값들을 불필요하게 List로 저장하던 부분 수정

jbilee committed:
fix(test): 테스트 패키지 오타 수정

작업 내용
- `validator` 패키지명 오타 수정

jbilee committed:
test(global): 로또 생성 관련 테스트 코드 수정

작업 내용
- `Lotto` 클래스 생성자 수정에 따라 불필요한 List 래퍼를 args 목록으로 대체

jbilee committed:
refactor(vendor): Vendor 생성자 리팩토링

작업 내용
- `Vendor` 클래스 생성자에 `RandomGenerator` 인스턴스를 받아 멤버 변수에 저장하도록 수정
- 로또 발행마다 `RandomGenerator` 인스턴스를 생성하던 부분 삭제

jbilee committed:
test(vendor): Vendor 도메인 테스트 수정

작업 내용
- `Vendor` 클래스 변동 사항에 따라 테스트 코드 수정
- 숫자 가독성을 위해 천 단위마다 `_` 구분자 추가

jbilee committed:
refactor(controller): 당첨 등수 관련 로직 수정

작업 내용
- `Scoreboard` 도메인에서 당첨 등수를 찾는 로직을 컨트롤러로 이동
- 컨트롤러에서 당첨 로또 번호와 보너스 번호를 입력 받는 로직을 별도 메소드로 분리
- `Scoreboard` 도메인 삭제

jbilee committed:
test(controller): 컨트롤러 테스트 작성

작업 내용
- `ScoreboardTest` 내 테스트를 `LottoControllerTest`로 이동
- `ScoreboardTest` 테스트 삭제

jbilee committed:
refactor(utility): RandomGenerator 로직 리팩토링

작업 내용
- 난수 List의 정렬을 `Lotto` 도메인에서도 하고 있어 `RandomGenerator`에서 정렬 기능 제외

jinyoungchoi95 commented on src/main/java/lotto/domain/Vendor.java:
```diff
@@ -10,10 +10,12 @@ public class Vendor {
 
     private static final int MAX_RANDOM_VALUE = 45;
     private static final int LOTTO_PRICE = 1000;
+    private final RandomGenerator randomGenerator;
```
> 3: 테스트 가능한 부분과 테스트하기 힘든 부분을 분리해 테스트 가능한 부분에 대해서만 단위 테스트를 진행한다.

현재 목적으로 하는 것은 domain 코드에 대해서 명확하게 테스트하는 것인데요. `VendorTest`에서 `issueLottos`에 대한 테스트를 하고는 있지만 정확한 테스트는 생성한 갯수를 테스트하는 것이 아닌 원하는 Lotto 객체를 만드는 것까지 테스트하는 것입니다.

현재는 원하는 갯수를 만든다로 우회했을 뿐이지 어떤 `Lotto` 객체가 생성되었는지는 테스트하지 못하는 상황은 똑같은거죠.

그래서 요구하는 내용은 `Random`에 대한 구현체를 어떻게 추상화하여 테스트할 수 있는 구조로 만드는 것이기 때문에 구현체를 추상화하는 것에 대해서 고민해보시면 좋을 것 같아요.

jbilee replied:
설명 감사합니다! 생각해보니 `Random`에 대한 구현체를 만들고, 제가 기대하는 `Lotto` 객체를 생성했는지를 1단계 피드백에서 말씀주신 `equals`를 재정의해 확인할 수 있을 것 같습니다.

jinyoungchoi95 commented on src/test/java/lotto/controller/LottoControllerTest.java:
```diff
@@ -0,0 +1,35 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class LottoControllerTest {
```
> Scoreboard가 수행하는 액션은 단순히 다른 도메인에게 당첨 등수를 찾아달라는 메세지를 던져 그 결과물을 합치기만 하는 것이며, 이는 컨트롤러에게 더 적합한 역할이라고 생각해 Scoreboard 도메인을 삭제하고 Scoreboard의 메소드를 컨트롤러가 수행하도록 수정했습니다.

말씀해주신 내용이 어떤 의도에서 말씀하신 것인지 이해는 가는데 테스트에 작성된 기능으로 보면 이건 도메인 로직처럼 보이지 않나요?

jbilee replied:
처음에는 코드의 형태(당첨 로또 객체의 `findWinningTier`만 호출하는 for문)만 보고 도메인 로직에 부합하기는 어려울 수 있겠다고 생각했는데 핵심 비즈니스 로직의 관점에서 보면 도메인 로직일 수 있을 것 같습니다.. 이 부분은 `List<Lotto>`를 저장하는 일급 컬렉션을 사용해 개선해볼까 합니다.

jinyoungchoi95 commented on src/main/java/lotto/utility/RandomGenerator.java:
위에도 말씀드렸지만 추상화에 대한 개념을 `RandomGenerator`에 녹여보세요.

jinyoungchoi95 commented on src/main/java/lotto/domain/Vendor.java:
```diff
@@ -28,15 +30,12 @@ public int calculateLottoCount() {
     }
 
     public List<Lotto> issueLottos() {
+        final int LOTTO_NUMBERS = 6;
```
지역변수는 컨벤션상 대문자로 표기하지 않습니다. 의미있는 상수로 변경하는 것이 맞겠네요.

jbilee replied:
지역변수의 컨벤션 확인했습니다, 감사합니다!

jinyoungchoi95 commented on src/main/java/lotto/domain/Vendor.java:
```diff
@@ -28,15 +30,12 @@ public int calculateLottoCount() {
 
     public List<Lotto> issueLottos() {
+        final int LOTTO_NUMBERS = 6;
         int lottoCount = this.calculateLottoCount();
```
`calculateLottoCount`가 public이 된 것은 테스트를 위한 public 메소드가 되는 것 아닐까요?

jbilee replied:
`calculateLottoCount`가 Vendor 내부에서만 쓰이니 private으로 하는게 맞는 것 같습니다. 테스트 코드를 다시 확인해보니 아래 두 테스트는 테스트하려는 성격이 같아 `calculateLottoCount`를 직접적으로 호출해보는 테스트를 삭제해도 `issueLottos`를 통해 간접적으로 확인이 될 것 같습니다.

```java
@DisplayName("구매 금액에 해당하는 로또 개수를 계산할 수 있다.")
@Test
void 구매_금액에_해당하는_로또_개수를_계산할_수_있다() {
    int expectedValue = 10;

    assertThat(vendor.calculateLottoCount()).isEqualTo(expectedValue);
}

@DisplayName("구매한 개수만큼의 로또를 생성할 수 있다.")
@Test
void 구매한_개수만큼의_로또를_생성할_수_있다() {
    int expectedCount = 10;
    List<Lotto> lottos = vendor.issueLottos();

    assertThat(lottos.size()).isEqualTo(expectedCount);
}
```

jinyoungchoi95 commented on src/test/java/lotto/domain/VendorTest.java:
```diff
@@ -14,19 +15,20 @@
     @BeforeEach
     void beforeEach() {
-        vendor = new Vendor(this.PURCHASE_AMOUNT);
+        vendor = new Vendor(this.randomGenerator, this.PURCHASE_AMOUNT);
```
junit5의 실행 객체 생성 단위는 메소드이기 때문에 `@BeforeEach`를 굳이 사용할 필요는 없습니다.

https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-instance-lifecycle

jbilee replied:
아하 그러면 `@BeforeEach` 대신 바로 필드로 생성하면 된다는 것인가요?

```java
// Before
class VendorTest {
    private final int PURCHASE_AMOUNT = 10_000;
    private Vendor vendor;

    @BeforeEach
    void beforeEach() {
        vendor = new Vendor(new RandomGenerator(), this.PURCHASE_AMOUNT);
    }
    
    // ...
}

// After
class VendorTest {
    private final int PURCHASE_AMOUNT = 10_000;
    private Vendor vendor = new Vendor(new RandomGenerator(), this.PURCHASE_AMOUNT);
}
```

충돌을 줄이기 위해 애초에 메소드마다 새 테스트 인스턴스가 생성되는군요 😳 소중한 자료 감사합니다!

jinyoungchoi95 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -22,18 +23,29 @@ public LottoController(InputView inputView, OutputView outputView) {
-
-        Scoreboard scoreboard = new Scoreboard();
-        List<WinningTier> winningTiers = scoreboard.findWinningTiers(lottos, winningLotto);
+        WinningLotto winningLotto = this.getWinningLotto();
+        List<WinningTier> winningTiers = this.getWinningTiers(lottos, winningLotto);
```
변수의 경우에는 파라메터와 네이밍이 동일할 수 있어 this를 붙이는 것이 의미가 있을 수 있으나 본인이 가진 메소드를 this를 붙이는 것은 코드 길이만 늘려 가독성을 더 떨어뜨릴 수 있습니다.

jbilee replied:
저는 오히려 가독성이 올라간다고 느꼈는데 오리의 말씀도 이해되는 것 같습니다! 제가 그동안 this 키워드에 많이 의존했을 수 있겠다는 생각이 듭니다. 😅

jinyoungchoi95 reviewed (changes requested):
안녕하세요 줄리 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

jbilee committed:
refactor(utility): RandomGenerator 로직 리팩토링

작업 내용
- `RandomNumberGenerator` 인터페이스 생성
- 직관적인 네이밍과 가독성을 위해 `RandomGenerator` 클래스명을 `LottoNumberGenerator`로 변경

jbilee committed:
refactor(lotto): Lotto 도메인 리팩토링

작업 내용
- Lotto 객체들끼리의 동일성을 비교할 수 있도록 `equals` 메소드 재정의

jbilee committed:
refactor(lotto): Lottos 일급 객체 생성

작업 내용
- `Lottos` 일급 객체를 생성하고 `LottoController`에서 전체 등수를 확인하는 로직을 `Lottos`로 이동
- `Vendor` 클래스에서 로또 발행 시 리스트 대신 `Lottos`를 반환하도록 수정

jbilee committed:
refactor(vendor): Vendor 클래스 로직 수정

작업 내용
- 클래스 내부에서만 쓰이는 `calculateLottoCount` 메소드를 private으로 전환

jbilee committed:
refactor(global): this 키워드 컨벤션 수정

작업 내용
- 인스턴스 메소드에 포함되어 있는 this 키워드 삭제
- 멤버 변수에는 this 키워드를 사용하도록 수정

jbilee committed:
refactor(view): 중복 코드 제거

작업 내용
- `OutputView` 메소드에서 `String.format`을 중복으로 사용하던 부분 제거

jbilee committed:
test(util): `LottoNumberGenerator` 테스트 코드 수정

작업 내용
- 테스트 클래스명을 `RandomGeneratorTest`에서 `LottoNumberGeneratorTest`로 변경
- Random API의 반환값에 대한 테스트 추가 (Random API를 사용해 무작위로 번호를 생성한다)

jbilee committed:
test(vendor): `Vendor` 클래스 테스트 코드 수정

작업 내용
- `@BeforeEach`를 삭제하고 `Vendor` 인스턴스를 멤버 변수에 저장
- private으로 변환된 메소드에 대한 테스트 삭제

jbilee committed:
test(vendor): `Vendor` 클래스 테스트 코드 추가

테스트 내용
- 로또 구매 시 무작위의 번호를 가진 로또 객체를 생성할 수 있다

jbilee committed:
test(lotto): `LottoControllerTest` 테스트 코드 이동

작업 내용
- `LottoControllerTest`의 테스트 코드를 `LottosTest` 도메인 테스트로 이동

jbilee committed:
refactor(vendor): `Vendor` 클래스 리팩토링

작업 내용
- 로또 번호 개수에 대한 지역변수를 상수로 전환

jbilee committed:
refactor(vendor): `Vendor` 클래스 리팩토링

작업 내용
- 멤버 변수명을 직관적으로 수정

jinyoungchoi95 commented on src/main/java/lotto/utility/LottoNumberGenerator.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.List;
+import java.util.Set;
+
+public class LottoNumberGenerator {
```
```java
    static class TestLottoNumberGenerator extends LottoNumberGenerator {
```
로 상속의 형태로 충분히 풀어 미션 요구사항을 완수한 것처럼 보일뿐  `RandomNumberGenerator`는 의미없는 추상화가 이루어져있고, 결국 도메인은 `LottoNumberGenerator`라는 구현체에 의존하고 있는 상황과 다름 없습니다.

```java
return generateNumbers(() -> (int) Math.ceil(Math.random() * maxNumber), count);
```

의도하는 바는, 랜덤한 동작을 하는 구현체 자체를 추상화하는 것입니다. 단순히 복잡하게 인터페이스를 주입하는 것이 아니라 `LottoNumberGenerator`자체를 추상화시키는거죠. 결과적으로 현재 미션의 목적은 랜덤한 행위를 어떻게 잘 추상화했고 그걸 어떻게 주입하여 풀어나가느냐이기때문에 아래 영상을 학습하시고 다음 미션진행에 참고하시면 좋을 것 같아요.

https://youtu.be/1VwRtsLDiyM?si=4lr2k8ZW6KYvn9lQ&t=498

jbilee replied:
영상 링크 감사합니다! 아직은 감이 정확히 안오지만 다른 자료와 함께 학습하면서 추상화를 익혀보겠습니다. 😀

jinyoungchoi95 reviewed (approved):
안녕하세요 줄리 😄

2단계 요구사항을 충분히 진행하신 것 같아 머지하겠습니다. 로또 미션기간동안 주고받았던 피드백이 앞으로도 많은 도움이 되길 바랍니다 😄

감사합니다 🙇‍♂️
